### PR TITLE
LogContext Push() and Clone()

### PR DESCRIPTION
**Pull request by nblumhardt**
This implements the suggestions in #733.

 * `Push()` is just a naming fix - `PushProperties()` actually pushes a collection of enrichers
 * `Clone()` lets worker threads opt-in to inheriting the parent's context stack

### CODE REVIEW

File: src/Serilog/Context/LogContext.cs [hunk 1]
```diff
@@ -75,37 +76,80 @@ public static class LogContext
         /// <returns>A token that must be disposed, in order, to pop properties back off the stack.</returns>
         public static IDisposable PushProperty(string name, object value, bool destructureObjects = false)
         {
+            return Push(new PropertyEnricher(name, value, destructureObjects));
+        }
+
+        /// <summary>
+        /// Push an enricher onto context, returning an <see cref="IDisposable"/>
```
Comment (adamchester) [inline src/Serilog/Context/LogContext.cs]: Grammar; could this be
> Push an enricher onto *the* context ?

File: src/Serilog/Context/LogContext.cs [hunk 2]
```diff
@@ -75,37 +76,80 @@ public static class LogContext
...
+        /// that must later be used to remove the property, along with any others that
+        /// may have been pushed on top of it and not yet popped. The property must
+        /// be popped from the same thread/logical call context.
+        /// </summary>
+        /// <param name="enricher">Log Properties to push onto the log context</param>
```
Comment (adamchester) [inline src/Serilog/Context/LogContext.cs]: why "Log Properties" (capitalised) ? I notice this was just copied, but maybe now is a good time to clean it up?

File: src/Serilog/Context/LogContext.cs [hunk 3]
```diff
@@ -75,37 +76,80 @@ public static class LogContext
...
 
             return bookmark;
         }
 
         /// <summary>
-        /// Push multiple properties onto the context, returning an <see cref="IDisposable"/>
-        /// that can later be used to remove the properties. The properties must
+        /// Push multiple enrichers onto the context, returning an <see cref="IDisposable"/>
+        /// that must later be used to remove the property, along with any others that
+        /// may have been pushed on top of it and not yet popped. The property must
```
Comment (adamchester) [inline src/Serilog/Context/LogContext.cs]: Regarding:
> along with any others that may have been pushed on top of it and not yet popped

Is this behaviour explained elsewhere? I think I may need to understand this better.

File: test/Serilog.Tests/Serilog.Tests.csproj [hunk 4]
```diff
@@ -29,7 +29,7 @@
     <DefineConstants>$(DefineConstants);APPDOMAIN;REMOTING;GETCURRENTMETHOD</DefineConstants>
   </PropertyGroup>
   <PropertyGroup Condition=" '$(TargetFramework)' == 'net46' ">
-    <DefineConstants>$(DefineConstants);ASYNCLOCAL</DefineConstants>
+    <DefineConstants>$(DefineConstants);ASYNCLOCAL;GETCURRENTMETHOD</DefineConstants>
```
Comment (adamchester) [inline test/Serilog.Tests/Serilog.Tests.csproj]: Was this change intended? Is there a reason behind it?

File: test/Serilog.Tests/Context/LogContextTests.cs [hunk 5]
```diff
@@ -25,6 +26,51 @@ public LogContextTests()
...
+                log.Write(Some.InformationEvent());
+                Assert.Equal(1, lastEvent.Properties["A"].LiteralValue());
+                Assert.Equal(2, lastEvent.Properties["B"].LiteralValue());
+                Assert.Equal(3, lastEvent.Properties["C"].LiteralValue());
+                Assert.Equal(4, lastEvent.Properties["D"].LiteralValue());
+            }
+        }
+
+        [Fact]
+        public void LogContextCanBeCloned()
```
Comment (adamchester) [inline test/Serilog.Tests/Context/LogContextTests.cs]: How does cloning work across threads and logical call contexts? Is there any sane way to test this?

File: src/Serilog/Context/LogContext.cs (hunk 1 above)
Comment (nblumhardt) [inline src/Serilog/Context/LogContext.cs]: Oops, good spotting :-)

File: src/Serilog/Context/LogContext.cs (hunk 3 above)
Comment (nblumhardt) [inline src/Serilog/Context/LogContext.cs]: The context is a stack, but each "push" produces the `IDisposable` bookmark that will pop the stack all the way to whatever state it was in prior to the Push(). This means:

```
x = Push();
Push();
Push();
x.Dispose();
```

will pop all three enrichers off the stack.

File: test/Serilog.Tests/Serilog.Tests.csproj (hunk 4 above)
Comment (nblumhardt) [inline test/Serilog.Tests/Serilog.Tests.csproj]: It lights up an additional test that we only run when `GetCurrentMethod()` is available (just noticed the inconsistency).

File: src/Serilog/Context/LogContext.cs (hunk 3 above)
Comment (adamchester) [inline src/Serilog/Context/LogContext.cs]: That clears it up; thanks.

File: test/Serilog.Tests/Context/LogContextTests.cs (hunk 5 above)
Comment (nblumhardt) [inline test/Serilog.Tests/Context/LogContextTests.cs]: Test added 👍

**Comment by nblumhardt** [comment]
Thanks for the review @adamchester. Should have all the feedback covered now 👍

**Comment by adamchester** [comment]
LGTM
